### DISCUSSION TITLE
text classification pipeline

### DISCUSSION THREAD

**Discussion by ahmed-moubtahij** [explosion/spaCy#8270] [Help: Coding & Implementations] [answered]
I have a list of sentences and I want to classify every one.

To achieve that, I want the `tok2vec` component to take every sentence and map it to its encoding from a pre-trained word2vec table (static vectors). Then, I want tok2vec's output to be the input for GRU as a sentence encoder. The output from that is to then be classified (binary). How can I achieve this architecture? None of the pipeline components seem to encode a sentence, nor give me the option to choose my encoder (lstm or other). The classification part goes with "textcat", but I'm not sure what I should put before that.

Here is my config file so far:
```
[paths]
train = null
dev = null
vectors = "w2v_table_processing/spacy_w2v_table"
init_tok2vec = null

[system]
seed = 0
gpu_allocator = null

[nlp]
lang = "en"
pipeline = ["tok2vec", "textcat"]
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
batch_size = 1000
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.tok2vec.model.embed]
include_static_vectors = true

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
gold_preproc = false
max_length = 0
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
gold_preproc = false
max_length = 0
limit = 0
augmenter = null

[training]
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
before_to_disk = null
logger = {"@loggers":"spacy.ConsoleLogger.v1"}

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]

[initialize]
vectors = ${paths.vectors}
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.components.textcat]
positive_label = "1"

[initialize.tokenizer]
```

**Comment by polm** [comment, marked as answer]
> Then, I want tok2vec's output to be the input for GRU as a sentence encoder. The output from that is to then be classified (binary). How can I achieve this architecture? None of the pipeline components seem to encode a sentence, nor give me the option to choose my encoder (lstm or other).

You can do this with spaCy but it really isn't designed to be used to explicitly specify neural architectures - it's a much higher level tool. If you want to design networks at this level you might want to look at [Thinc](https://thinc.ai), which is used to implement spaCy's models. If you have a custom architecture, you can implement it in Thinc and then make a spaCy component (you can also use Pytorch or anything). 

I would strongly encourage you to use [the textcat quickstart](https://spacy.io/usage/training#quickstart) as-is and look at [the example textcat projects](https://github.com/explosion/projects/tree/v3/tutorials) to get an idea of how spaCy is designed to be used and see if it matches your approach and works on your project. If you already have training data getting a working model from the quickstart config should be very fast. 

> None of the pipeline components seem to encode a sentence,

You can get an embedding for any doc or span using the `.vector` attribute ([docs](https://spacy.io/api/doc#vector)). It defaults to an average of token vectors, but you can set a user hook to have it do anything you want. (It's true this isn't a pipeline component.)

**Reply by svlandeg** [reply]
Just wanted to add that a good place to start to implement custom trainable components, is here: https://spacy.io/usage/layers-architectures#components. That page also has more information about Thinc, or using e.g. a PyTorch model in a spaCy pipeline component.

**Reply by ahmed-moubtahij** [reply]
Complete answer, I'll be studying this for a few. Thank you polm and svlandeg.
